LibOS/shim/src/ipc/shim_ipc.c

```diff
@@ -368,11 +368,20 @@ int send_ipc_message (struct shim_ipc_msg * msg, struct shim_ipc_port * port)
     assert(msg->size >= IPC_MSG_MINIMAL_SIZE);
     msg->src = cur_process.vmid;
 
-    int ret = DkStreamWrite(port->pal_handle, 0, msg->size, msg, NULL);
+    debug("send ipc message to port %p (handle %p)\n", port,
+          port->pal_handle);
+
+    PAL_HANDLE pal_handle = port->pal_handle;
+    rmb();
```

Comment:
Why is a barrier needed here?  This also merits a comment.

Reply:
I added some comment here, as well as in `shim_ipc_helper.c` where `port->deleted` is set to true. Basically we need the barrier here to ensure `port->deleted` is up-to-date in the memory.